x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/SourceOnlySnapshotRepository.java

```diff
@@ -104,15 +105,18 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
     }
 
     @Override
-    public void snapshotShard(IndexShard shard, Store store, SnapshotId snapshotId, IndexId indexId, IndexCommit snapshotIndexCommit,
-                              IndexShardSnapshotStatus snapshotStatus) {
-        if (shard.mapperService().documentMapper() != null // if there is no mapping this is null
-            && shard.mapperService().documentMapper().sourceMapper().isComplete() == false) {
+    public void snapshotShard(Store store, MapperService mapperService, SnapshotId snapshotId, IndexId indexId,
+                              IndexCommit snapshotIndexCommit, IndexShardSnapshotStatus snapshotStatus) {
+        if (mapperService.documentMapper() != null // if there is no mapping this is null
+            && mapperService.documentMapper().sourceMapper().isComplete() == false) {
             throw new IllegalStateException("Can't snapshot _source only on an index that has incomplete source ie. has _source disabled " +
                 "or filters the source");
         }
-        ShardPath shardPath = shard.shardPath();
-        Path dataPath = shardPath.getDataPath();
+        Directory unwrap = FilterDirectory.unwrap(store.directory());
+        if (unwrap instanceof FSDirectory == false) {
```

Comment:
Maybe add an `assert unwrap instanceof FSDirectory`? We have a number of tests that allow for an exception in this code and it might not be the worst idea to lock this down for good?
Also, with that, maybe just let it die on a `ClassCastException` below, that's probably good enough for debugging?